### PR TITLE
Create SBOM using Syft

### PR DESCRIPTION
**Pull request by Brend-Smits**
The SBOM is created using tooling from Anchore, called [Syft](https://github.com/anchore/syft). Syft was chosen because of the ease of use and the good integration with our tech stack. 
I think this could have been a separate step in the workflow, but in the end, I chose not to go this route. The reason being is that it will be a lot of repeating the same code and will raise a new problem that we have to upload a temporary artifact or pass along a different output (which may have issues being passed around since the SBOM's can be relatively large in size).

This PR also removes the filter sbom spdx.

### CODE REVIEW

File: .goreleaser.draft.yml [hunk 1]
```diff
@@ -39,7 +39,8 @@ archives:
 sboms:
   - artifacts: archive
     documents:
-      - "{{ .ProjectName }}_{{ .Version }}_sbom.spdx.json"
+      - "binary":   ["{{ .Binary }}_{{ .Version }}_{{ .Os }}_{{ .Arch }}.sbom"]
```
Comment (marcofranssen) [inline .goreleaser.draft.yml]:
```suggestion
      - binary:   ["{{ .Binary }}_{{ .Version }}_{{ .Os }}_{{ .Arch }}.sbom"]
```

Also ensure to apply same on the `.goreleaser.yml`. Preferably keep them in sync in the same commit.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -81,6 +81,8 @@ jobs:
         with:
           cosign-release: 'v1.4.1'
 
+      - uses: anchore/sbom-action/download-syft@v0.6.0
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:84]:
```suggestion
      - name: Install syft
         uses: anchore/sbom-action/download-syft@v0.6.0
```

**Review by marcofranssen — CHANGES_REQUESTED** [review]

File: .goreleaser.draft.yml (hunk 1 above)
Comment (Brend-Smits) [reply .goreleaser.draft.yml]: I think you reviewed, just when I pushed a new commit fixing this. Can you re-review?

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (Brend-Smits) [reply .github/workflows/ci.yaml]: Thanks!

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -139,6 +142,73 @@ jobs:
...
+        run: |
+          echo '${{ secrets.COSIGN_PRIVATE_KEY }}' > cosign.key
+
+      - name: Create SBOM from container with Syft
+        run: |
+          syft ghcr.io/philips-labs/slsa-provenance:${{ needs.release.outputs.container_digest }} -o spdx-json > sbom.spdx.json
+
+      - name: Upload Container SBOM to OCI
+        run: |
+          cosign upload blob -f sbom.spdx.json ${{ matrix.repo }}/attestations@${{ needs.release.outputs.container_digest }}
```
Comment (JeroenKnoops) [inline .github/workflows/ci.yaml:189]: Why are we uploading this as a `blob`? We can also can use the attestation type `spdx`: 

```
cosign attest --predicate sbom-spdx.json --type spdx ........
```

**Review by JeroenKnoops — APPROVED** [review]

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -139,6 +142,73 @@ jobs:
...
+    steps:
+      - name: Set up Go
+        uses: actions/setup-go@v2.1.5
+        with:
+          go-version: 1.17
+
+      - name: Install cosign
+        uses: sigstore/cosign-installer@v1.4.1
+        with:
+          cosign-release: 'v1.4.1'
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:164]:
```suggestion
          cosign-release: 'v1.5.1'
```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -139,6 +142,73 @@ jobs:
...
+        uses: anchore/sbom-action/download-syft@v0.6.2-draft
+
+      - name: Checkout
+        uses: actions/checkout@v2.4.0
+        with:
+          fetch-depth: 0
+
+      - name: Login to Container registries
+        run: |
+          echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u philipssoftware --password-stdin
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:176]: I think we have a matrix per registry, so we should only login to the registry for this particular matrix.

As we are then only logging in on one registry at a time we might reuse the docker login action.

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -139,6 +142,73 @@ jobs:
...
+        run: |
+          cosign sign --key cosign.key ${{ matrix.repo }}/attestations@${{ needs.release.outputs.container_digest }}
+        env:
+          COSIGN_PASSWORD: ${{ secrets.COSIGN_PASSWORD }}
+
+      - name: Verify the signature of the container SBOM
+        run: |
+          cosign verify --key cosign.pub ${{ matrix.repo }}/attestations@${{ needs.release.outputs.container_digest }}
+
+      - name: Logout from Container registries
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:201]: See login comment above

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -139,6 +142,73 @@ jobs:
...
+        run: |
+          syft ghcr.io/philips-labs/slsa-provenance:${{ needs.release.outputs.container_digest }} -o spdx-json > sbom.spdx.json
+
+      - name: Upload Container SBOM to OCI
+        run: |
+          cosign upload blob -f sbom.spdx.json ${{ matrix.repo }}/attestations@${{ needs.release.outputs.container_digest }}
+
+      - name: Sign the container SBOM
+        run: |
+          cosign sign --key cosign.key ${{ matrix.repo }}/attestations@${{ needs.release.outputs.container_digest }}
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:193]: We should not apply this workarround on our opensource repos. Simply go with @JeroenKnoops his suggestion on using `cosign attest`

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -139,6 +142,73 @@ jobs:
...
+        env:
+          COSIGN_PASSWORD: ${{ secrets.COSIGN_PASSWORD }}
+
+      - name: Verify the signature of the container SBOM
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:197]: This step can also be removed when we use `cosign attest`

File: .goreleaser.draft.yml [hunk 9]
```diff
@@ -36,6 +36,15 @@ archives:
       - goos: windows
         format: zip
 
+sboms:
+  -
```
Comment (marcofranssen) [inline .goreleaser.draft.yml:40]:
```suggestion
```

File: .goreleaser.yml [hunk 10]
```diff
@@ -36,6 +36,15 @@ archives:
       - goos: windows
         format: zip
 
+sboms:
+  -
```
Comment (marcofranssen) [inline .goreleaser.yml:40]:
```suggestion
```

File: .goreleaser.yml [hunk 11]
```diff
@@ -36,6 +36,15 @@ archives:
...
+    id: archive-sbom
```
Comment (marcofranssen) [inline .goreleaser.yml:41]:
```suggestion
  - id: archive-sbom
```

**Review by marcofranssen — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -141,6 +144,51 @@ jobs:
...
+          echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u philipssoftware --password-stdin
+          echo "${{ secrets.GITHUB_TOKEN }}" | docker login -u ${{ github.actor }} --password-stdin ghcr.io
+
+      - name: Attach SBOM
+        run: |
+          echo '${{ secrets.COSIGN_PUBLIC_KEY }}' > cosign.pub
+          echo '${{ secrets.COSIGN_PRIVATE_KEY }}' > cosign.key
+          IFS=,
+          for t in ${TAGS}; do
+            cosign verify --key cosign.pub ${{ matrix.repo }}:${{ needs.release.outputs.container_digest }}
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:179]:
```suggestion
            cosign verify --key cosign.pub "${{ matrix.repo }}:${t}"
```

File: .github/workflows/ci.yaml [hunk 13]
```diff
@@ -141,6 +144,51 @@ jobs:
...
+            syft ${{ matrix.repo }}:${{ needs.release.outputs.container_digest }} -o spdx-json > sbom-spdx.json
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:180]:
```suggestion
            syft "${{ matrix.repo }}:${t}" -o spdx-json > sbom-spdx.json
```

File: .github/workflows/ci.yaml [hunk 14]
```diff
@@ -141,6 +144,51 @@ jobs:
...
+            cosign attest --predicate sbom-spdx.json --type spdx --key cosign.key ${{ matrix.repo }}@${{ needs.release.outputs.container_digest }}
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:181]:
```suggestion
            cosign attest --predicate sbom-spdx.json --type spdx --key cosign.key "${{ matrix.repo }}:${t}"
```

File: .github/workflows/ci.yaml [hunk 15]
```diff
@@ -141,6 +144,51 @@ jobs:
...
+            cosign verify-attestation --key cosign.pub ${{ matrix.repo }}@${{ needs.release.outputs.container_digest }}
```
Comment (marcofranssen) [inline .github/workflows/ci.yaml:182]:
```suggestion
            cosign verify-attestation --key cosign.pub "${{ matrix.repo }}:${t}"
```

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Looks good, still would like to check using crane and local commandline from a created draft release using the current state of the PR all this sboms and things are accordingly released to ensure also the goreleaser configs are accurate.

File: .github/workflows/ci.yaml (hunk 15 above)
Comment (Brend-Smits) [reply .github/workflows/ci.yaml]: Nice!

**Comment by Brend-Smits** [comment]
https://github.com/philips-labs/slsa-provenance-action/runs/5235719140?check_suite_focus=true
Feel free to test with Crane @marcofranssen 
It's ready for merge. On merge, please Squash the commits.

**Pull request merged**